content/en/blog/_posts/2024-03-19-go-workspaces.md

```diff
@@ -0,0 +1,186 @@
+the same time, Kubernetes got more and more popular. For many people,
+Kubernetes was their first experience with Go, and it wasn't always a good
+experience.
+
+Our eccentricities also impacted people who consumed some of our code, such as
+our client library and the code-generation tools (which turned out to be useful
+in the growing ecosystem of custom resources). The tools only worked if you
+stored your code in a particular GOPATH-compatible directory structure, even
+though GOPATH had been replaced by modules more than 4 years before.
+
+This state persisted because of the confluence of 3 factors:
+  1) Most of the time it only hurt a little (punctuated with short moments of
+     more acute pain).
+  2) Kubernetes was still growing in popularity - we all had other, more urgent
+     things to work on.
+  3) The fix was not obvious, and whatever we came up with was going to be both
+     hard and tedious.
+
+As a Kubernetes maintainer and long-timer, my fingerprints were all over the
+build system, the code-generation tools, and the `hack` scripts. While the pain
+of our mess may have been low _on_average_, I was one of the people who felt it
+regularly.
+
+## Enter workspaces
+
+Along the way, the Go language team saw what we (and others) were doing and
+didn't love it. They designed a new way of stitching multiple modules together
+into a new "workspace" concept. Once enrolled in a workspace, Go tools had
+enough information to work in any directory structure and across modules,
+without GOPATH or symlinks or other dirty tricks.
+
+When I first saw this proposal I knew that this was the way out. This was how
+to break the logjam. If workspaces was the technical solution, then I would
+put in the work to make it happen.
+
+## The work
+
+Adopting workspaces was deceptively easy. I very quickly had the codebase
+compiling and running tests with workspaces enabled. I set out to purge the
+repository of anything GOPATH related. That's when I hit the first real bump -
+the code-generation tools.
+
+We had about a dozen tools, totalling several thousand lines of code. All of
+them were built using an internal framework called
+[gengo](https://github.com/kubernetes/gengo), which was built on Go's own
+parsing libraries. There were 2 main problems:
+
+  1) Those parsing libraries didn't understand modules or workspaces.
+  2) GOPATH allowed us to pretend that Go "package paths" and directories on
+     disk were interchangeable in trivial ways. They are not.
+
+Switching to a [modules- and workspaces-aware parsing](https://pkg.go.dev/golang.org/x/tools@v0.19.0/go/packages)
+library was the first step. Then I had to make a long series of changes to
+each of the code-generation tools. Critically, I had to find a way to do it
+that was possible for some other person to review! After some conversations
+with likely reviewers, we agreed that a series of small commits would be
```

Comment:
I think it could be interesting for the community to understand who you knew to approach. I know your long history in the project means you have a lot of historical knowlege/context to lean on, but it might be worth mentioning the SIGs you approached, or individuals, and why they would be suitable (just a couple of lines will do it)

Reply:
Added, PTAL at next push.